File: lib/allocator.js
const LBS_TO_KG = 0.453592;
const AC_TO_HA = 0.404685642;

function computeBushelsNeeded(rate, area, unitSystem, lbPerBu) {
  const kgPerBu = lbPerBu * LBS_TO_KG;
  if (unitSystem === 'imperial') {
    return (rate * area) / lbPerBu;
  }
  return (rate * area) / kgPerBu;
}

function buildTankList(caps) {
  const labels = ['Tank 1', 'Tank 2', 'Tank 3', 'Tank 4', 'Saddle'];
  const tanks = caps.map((c, i) => ({ label: labels[i] || `Tank ${i+1}`, capacity: c, index: i }));
  tanks.sort((a, b) => b.capacity - a.capacity);
  return tanks;
}

function allocate(modelCaps, products, unitSystem) {
  // products: [{ name, rate, lbPerBu }]
  if (!modelCaps || !products || products.length === 0) return { errors: ['Invalid input'] };

  const productsBu = products.map(p => ({
    name: p.name,
    rate: p.rate,
    bu: computeBushelsNeeded(p.rate, p.area, unitSystem, p.lbPerBu)
  }));

  productsBu.sort((a, b) => b.bu - a.bu);

  const tanks = buildTankList(modelCaps);
  const plan = [];
  let tankPtr = 0;

  for (const prod of productsBu) {
    let remaining = prod.bu;
    while (remaining > 0 && tankPtr < tanks.length) {
      const t = tanks[tankPtr];
      const fill = Math.min(t.capacity, remaining);
      plan.push({ tankLabel: t.label, capacity: t.capacity, product: prod.name, fillBu: fill, fillPct: (fill / t.capacity) * 100 });
      remaining -= fill;
      tankPtr++;
    }
    if (remaining > 0) {
      plan.push({ tankLabel: 'UNALLOCATED', capacity: 0, product: prod.name, fillBu: remaining, fillPct: 0 });
    }
  }

  const totalCapacity = modelCaps.reduce((a, b) => a + b, 0);
  const totalBuNeeded = productsBu.reduce((a, b) => a + b.bu, 0);
  const utilizationPct = Math.min(100, (totalBuNeeded / totalCapacity) * 100);

  return { productsBu, plan, utilizationPct };
}

module.exports = { allocate, computeBushelsNeeded, buildTankList };

File: iOS/BourgaultTankOptimizer/Allocator.swift
//
//  Allocator.swift
//  BourgaultTankOptimizer
//
//  Core allocation algorithm for distributing products across tanks
//

import Foundation

/// Handles tank allocation calculations
class Allocator {
    
    // Conversion constants
    private static let lbsToKg = 0.453592
    private static let acresToHa = 0.404685642
    
    /// Compute bushels needed for a product
    /// - Parameters:
    ///   - rate: Application rate (lb/ac or kg/ha depending on unit system)
    ///   - area: Field area (acres or hectares depending on unit system)
    ///   - unitSystem: Current unit system
    ///   - densityLbPerBu: Product density in pounds per bushel
    /// - Returns: Bushels required
    static func computeBushelsNeeded(rate: Double, area: Double, unitSystem: UnitSystem, densityLbPerBu: Double) -> Double {
        let densityKgPerBu = densityLbPerBu * lbsToKg
        
        switch unitSystem {
        case .imperial:
            // bushels = (lb/ac * ac) / (lb/bu)
            return (rate * area) / densityLbPerBu
        case .metric:
            // bushels = (kg/ha * ha) / (kg/bu)
            return (rate * area) / densityKgPerBu
        }
    }
    
    /// Build sorted tank list from capacities
    /// - Parameter capacities: Array of tank capacities in bushels
    /// - Returns: Array of tanks sorted by capacity (largest first)
    static func buildTankList(capacities: [Double]) -> [Tank] {
        let labels = ["Tank 1", "Tank 2", "Tank 3", "Tank 4", "Saddle"]
        let tanks = capacities.enumerated().map { index, capacity in
            Tank(
                label: index < labels.count ? labels[index] : "Tank \(index + 1)",
                capacity: capacity,
                index: index
            )
        }
        return tanks.sorted { $0.capacity > $1.capacity }
    }
    
    /// Allocate products to tanks using a greedy algorithm
    /// - Parameters:
    ///   - model: Tank model with capacities
    ///   - selections: Product selections with rates
    ///   - fieldSize: Field size in current unit system
    ///   - unitSystem: Current unit system
    /// - Returns: Allocation result with tank assignments
    static func allocate(model: TankModel, selections: [ProductSelection], fieldSize: Double, unitSystem: UnitSystem) -> AllocationResult {
        
        // Validate inputs
        guard fieldSize > 0, !selections.isEmpty else {
            return AllocationResult(
                model: model,
                fieldSize: fieldSize,
                unitSystem: unitSystem,
                productsBushels: [],
                tankAllocations: [],
                utilizationPercent: 0,
                errors: ["Enter a valid field size and at least one product."]
            )
        }
        
        // Calculate bushels needed for each product
        var productsBushels: [(name: String, rate: Double, bushels: Double)] = []
        for selection in selections {
            let bushels = computeBushelsNeeded(
                rate: selection.rate,
                area: fieldSize,
                unitSystem: unitSystem,
                densityLbPerBu: selection.product.densityLbPerBu
            )
            productsBushels.append((
                name: selection.product.name,
                rate: selection.rate,
                bushels: bushels
            ))
        }
        
        // Sort products by bushels needed (largest first) for greedy allocation
        productsBushels.sort { $0.bushels > $1.bushels }
        
        // Build sorted tank list
        let tanks = buildTankList(capacities: model.capacities)
        
        // Greedy allocation algorithm
        var allocations: [TankAllocation] = []
        var tankIndex = 0
        
        for productInfo in productsBushels {
            var remaining = productInfo.bushels
            
            // Fill tanks with this product until it's all allocated or tanks run out
            while remaining > 0 && tankIndex < tanks.count {
                let tank = tanks[tankIndex]
                let fill = min(tank.capacity, remaining)
                
                allocations.append(TankAllocation(
                    tankLabel: tank.label,
                    capacity: tank.capacity,
                    productName: productInfo.name,
                    fillBushels: fill,
                    fillPercent: (fill / tank.capacity) * 100
                ))
                
                remaining -= fill
                tankIndex += 1
            }
            
            // If product couldn't be fully allocated, mark as unallocated
            if remaining > 0 {
                allocations.append(TankAllocation(
                    tankLabel: "UNALLOCATED",
                    capacity: 0,
                    productName: productInfo.name,
                    fillBushels: remaining,
                    fillPercent: 0
                ))
            }
        }
        
        // Calculate utilization
        let totalCapacity = model.totalCapacity
        let totalBushelsNeeded = productsBushels.reduce(0) { $0 + $1.bushels }
        let utilizationPercent = min(100, (totalBushelsNeeded / totalCapacity) * 100)
        
        return AllocationResult(
            model: model,
            fieldSize: fieldSize,
            unitSystem: unitSystem,
            productsBushels: productsBushels,
            tankAllocations: allocations,
            utilizationPercent: utilizationPercent,
            errors: []
        )
    }
}
